unison-share-api/src/Unison/Server/Backend.hs

```diff
@@ -962,36 +903,124 @@ prettyDefinitionsForHQName path mayRoot renderWidth suffixifyBindings rt codebas
       renderedMisses
 
-renderDoc ::
+mkTypeDefinition ::
+  Codebase IO Symbol Ann ->
+  PPED.PrettyPrintEnvDecl ->
+  Path.Path ->
+  V2Branch.CausalBranch Sqlite.Transaction ->
+  Width ->
+  Reference ->
+  [(HashQualifiedName, UnisonHash, Doc.Doc)] ->
+  DisplayObject
+    (AnnotatedText (UST.Element Reference))
+    (AnnotatedText (UST.Element Reference)) ->
+  Backend IO TypeDefinition
+mkTypeDefinition codebase pped namesRoot rootCausal width r docs tp = do
+  let bn = bestNameForType @Symbol (PPED.suffixifiedPPE pped) width r
+  tag <-
+    liftIO $ Codebase.runTransaction codebase do
+      causalAtPath <- Codebase.getShallowCausalAtPath namesRoot (Just rootCausal)
+      branchAtPath <- V2Causal.value causalAtPath
+      typeEntryTag <$> typeListEntry codebase branchAtPath (ExactName (NameSegment bn) r)
+  pure $
+    TypeDefinition
+      (HQ'.toText <$> PPE.allTypeNames fqnPPE r)
+      bn
+      tag
+      (bimap mungeSyntaxText mungeSyntaxText tp)
+      docs
+  where
+    fqnPPE = PPED.unsuffixifiedPPE pped
+
+mkTermDefinition ::
+  Codebase IO Symbol Ann ->
   PPED.PrettyPrintEnvDecl ->
+  Path.Path ->
+  V2Branch.CausalBranch Sqlite.Transaction ->
   Width ->
+  Reference ->
+  [(HashQualifiedName, UnisonHash, Doc.Doc)] ->
+  DisplayObject
+    (AnnotatedText (UST.Element Reference))
+    (AnnotatedText (UST.Element Reference)) ->
+  Backend IO TermDefinition
+mkTermDefinition codebase termPPED namesRoot rootCausal width r docs tm = do
+  let referent = Referent.Ref r
+  (ts, branchAtPath) <- liftIO $ Codebase.runTransaction codebase do
+    ts <- Codebase.getTypeOfTerm codebase r
+    causalAtPath <- Codebase.getShallowCausalAtPath namesRoot (Just rootCausal)
+    branchAtPath <- V2Causal.value causalAtPath
+    pure (ts, branchAtPath)
+  let bn = bestNameForTerm @Symbol (PPED.suffixifiedPPE termPPED) width (Referent.Ref r)
+  tag <-
+    lift
+      ( termEntryTag
+          <$> termListEntry codebase branchAtPath (ExactName (NameSegment bn) (Cv.referent1to2 referent))
+      )
+  mk ts bn tag
+  where
+    fqnTermPPE = PPED.unsuffixifiedPPE termPPED
+    mk Nothing _ _ = throwError $ MissingSignatureForTerm r
+    mk (Just typeSig) bn tag = do
+      -- We don't ever display individual constructors (they're shown as part of their
+      -- type), so term references are never constructors.
+      let referent = Referent.Ref r
+      pure $
+        TermDefinition
+          (HQ'.toText <$> PPE.allTermNames fqnTermPPE referent)
+          bn
+          tag
+          (bimap mungeSyntaxText mungeSyntaxText tm)
+          (formatSuffixedType termPPED width typeSig)
+          docs
+
+-- | Given an arbitrary query and perspective, find the project root the query belongs in,
+-- then return that root and the query relocated to that project root.
+relocateToProjectRoot :: Path -> HQ.HashQualified Name -> V2Branch.Branch Sqlite.Transaction -> Sqlite.Transaction (Either BackendError (Path, HQ.HashQualified Name))
```

Comment:
This is new.

When working with project roots in loose code we need to ensure that we re-write the query when we adjust the perspective.

E.g. if your query and perspective are (`map`, `base.main.data.List`) we rewrite it to (`data.List.map`, `base.main`) so it's at the project level.

This gives better results in pretty much every situation since otherwise you can end up printing a lot of hashes otherwise. It also means things are suffixified at the project level which helps with clarity and consistency as you navigate around.